bigtable/client/row_reader_test.cc

```diff
@@ -619,3 +625,51 @@ TEST_F(RowReaderTest, RowReaderConstructorDoesNotCallRpc) {
       bigtable::Filter::PassAllFilter(), std::move(retry_policy_),
       std::move(backoff_policy_), std::move(parser_factory_));
 }
+
+TEST_F(RowReaderTest, FailedStreamRetryNewContext) {
+  using namespace ::testing;
+  // Every retry should use a new ClientContext object.
+  auto* stream = new MockResponseStream();  // wrapped in unique_ptr by ReadRows
+  auto parser = absl::make_unique<ReadRowsParserMock>();
+  parser->SetRows({"r1"});
+
+  void* previous_context = nullptr;
+  EXPECT_CALL(*retry_policy_, setup_impl(_))
+      .Times(2)
+      .WillRepeatedly(Invoke([&previous_context](grpc::ClientContext& context) {
+        // This is a big hack, we want to make sure the context is new,
```

Comment:
Thanks for calling it out in a comment, but isn't this potentially flaky? Some compiler could call delete before new, and the allocator is permitted to reuse the memory?

```

TEST(ATest, ReplacePtr) {
  auto a = absl::make_unique<int>();
  auto p = a.get();
  for (int i = 0; i < 1000; i++) {
    a = absl::make_unique<int>();
    EXPECT_NE(p, a.get());
  }
}
```
leads in my compiler to:
```
Expected: (p) != (a.get()), actual: 0x7fa401c027a0 vs 0x7fa401c027a0
Segmentation fault: 11
```


On the flip side, looks like GRPC handles this by cancelling the second RPC: https://github.com/grpc/grpc/pull/2775/files
so the integration tests should catch this.

Reply:
It is brittle, because it depends on the implementation, but not likely to be flaky: the new context is created first and then the old context is deleted.  If we ever changed the implementation then it could get flaky, yes.

I tried other approaches, e.g., trying to see if `c_call()` works as a "used" marker in the `ClientContext`.  None worked for me :disappointed: